streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -559,47 +565,34 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
     catchAllCause(pf.applyOrElse[Cause[E], ZStream[R1, E1, O1]](_, ZStream.halt(_)))
 
   /**
-   * Re-chunks the elements of the stream into chunks of
-   * `n` elements each.
-   * The last chunk might contain less than `n` elements
+   * Re-chunks the chunks of the stream into non-empty chunks of size bounded by `max`.
    */
-  def chunkN(n: Int): ZStream[R, E, O] = {
-    case class State[X](buffer: Chunk[X], done: Boolean)
+  def chunkLimit(max: Int): ZStream[R, E, Chunk[O]] =
+    aggregateChunks(ZTransducer.chunkLimit(max))
 
-    def emitOrAccumulate(
-      buffer: Chunk[O],
-      done: Boolean,
-      ref: Ref[State[O]],
-      pull: ZIO[R, Option[E], Chunk[O]]
-    ): ZIO[R, Option[E], Chunk[O]] =
-      if (buffer.size < n) {
-        if (done) {
-          if (buffer.isEmpty)
-            Pull.end
-          else
-            ref.set(State(Chunk.empty, true)) *> Pull.emit(buffer)
-        } else
-          pull.foldM({
-            case Some(e) => Pull.fail(e)
-            case None    => emitOrAccumulate(buffer, true, ref, pull)
-          }, ch => emitOrAccumulate(buffer ++ ch, false, ref, pull))
-      } else {
-        val (chunk, leftover) = buffer.splitAt(n)
-        ref.set(State(leftover, done)) *> Pull.emit(chunk)
-      }
+  /**
+   * Re-chunks the chunks of the stream into non-empty chunks of a fixed `size`.
+   * The last chunk may have less than `size` elements.
+   */
+  def chunkN(size: Int): ZStream[R, E, Chunk[O]] =
```

Comment:
Sorry, missed this: the signature here changed. Should stay `ZStream[R, E, O]`. Same for `chunkLimit` which is newly added but should also use `ZStream[R, E, O]`.

Reply:
@iravid Wouldn't the rationale for chunking be to subsequently operate on chunks?
The current signature allows for
```scala
byteStream.chunkN(4).map(_.toByteBuffer.readInt)
```
If we change the signature, we would have to write this as
```scala
byteStream.chunkN(4).mapChunks(_.toByteBuffer.readInt)
```

Reply:
Ah I see where the confusion stems from. 

No, the purpose of the chunking combinators on ZStream is to modify the underlying chunking structure. For most cases, it is unnecessary to expose those chunks because all operators work on chunks automatically.

The chunks can be operated on as you suggested with `mapChunks`, or by exposing them using the `ZStream` version of your `ZTransducer.chunks` transducer that we'll add soon.

Reply:
Ah yes. `flattenChunks` is too eager. Let's fix that in a follow-up using `BufferedPull` instead of `mapConcatChunk`.

Reply:
@iravid 
Just sharing my thoughts.

> No, the purpose of the chunking combinators on ZStream is to modify the underlying chunking structure. For most cases, it is unnecessary to expose those chunks because all operators work on chunks automatically.

I agree with the first part. And the new signatures make this explicit!

Are there any combinators, other than `chunkN`, that alter the chunking structure?

Is there a usecase that alters the chunking structure and then subsequently operates on individual elements? I don't consider the above testcase a valid use (it just requires input arriving in chunks).

>The chunks can be operated on as you suggested with mapChunks, or by exposing them using the ZStream version of your ZTransducer.chunks transducer that we'll add soon.

Yes, but this penalizes users who would want to operate on chunks with a redundant `un-chunk + re-chunk`.

Reply:
Alternatively, we can remove the `chunk*` combinators on `ZStream` and let users construct the argument for `aggregateChunks`.

For my usecase, I only need `chunkLimit` on `ZTransducer`. It really doesn't make sense on `ZStream`, IMO.